fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/security/Base64URLData.java

```diff
@@ -0,0 +1,48 @@
+package com.github.dedis.popstellar.model.objects.security;
+
+import java.util.Arrays;
+import java.util.Base64;
+
+/** Represents a data that can be encoded into a Base64 form */
+public class Base64URLData {
+
+  protected final byte[] data;
+
+  public Base64URLData(byte[] data) {
+    this.data = data;
+  }
+
+  public Base64URLData(String data) {
+    this(decode(data));
+  }
+
+  public byte[] getData() {
+    return Arrays.copyOf(data, data.length);
+  }
+
+  /** @return the Base64 - encoded string representation of the data */
+  public String getEncoded() {
+    return encode(data);
+  }
+
+  private static byte[] decode(String data) {
+    return Base64.getUrlDecoder().decode(data);
+  }
+
+  private static String encode(byte[] data) {
+    return Base64.getUrlEncoder().encodeToString(data);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (!(o instanceof Base64URLData)) return false;
```

Comment:
There are warnings on sonarcloud about subclasses don't override the equals.
And the equals of Base64URLData use "instanceof Base64URLData" instead of the default "getClass() != o.getClass()", is it intentional so that all subclass equality are only based on the byte[] data ?
For example this code will return that they are equals.
```
Base64URLData base64URLData = new Base64URLData(VALID_PRIVATE_KEY);
PlainPrivateKey plainPrivateKey = new PlainPrivateKey(VALID_PRIVATE_KEY);
assertEquals(base64URLData, plainPrivateKey);
```

Reply:
Yes, this is intended.

The sonarcloud issue is a false positive because the fields are not supposed to be tested for equality.

And I designed this so that it compares the data even if the types are not exactly the same.

If you think this is a bad idea, just let me know.

Reply:
By type-aware equality I meant that in a test you wouldn't have been able to assert that `PublicKey(x) != PlainPrivateKey(x)`

PS: you needed to go on the outdated commit to be able to respond directly 😄

Reply:
OK, I intended that at first, but it doesn't make sense now that I think about it more. I fixed it in the latest commits.